packages/next/src/server/lib/trace/tracer.ts

```diff
@@ -203,15 +208,21 @@ class NextTracerImpl implements NextTracer {
 
         const result = fn(span)
 
-        if (isPromise(result)) {
-          result.then(
-            () => span.end(),
-            (err) => closeSpanWithError(span, err)
-          )
-        } else {
+        if (!isPromise(result)) {
           span.end()
+          return result
         }
 
+        result
+          .then((value) => {
+            span.end()
+            return value
+          })
+          .catch((err) => {
+            closeSpanWithError(span, err)
+            throw err
+          })
+
         return result
```

Comment:
For some reason, when I return the result with `then` and `catch`  it will start breaking tests because it will cause force reloads in dev mode.

I wasn't able to find the cause, so I'll leave it as it was before.